src/app/shared/services/dynamic-data/dynamic-data.service.ts

```diff
@@ -178,9 +178,17 @@ export class DynamicDataService extends AsyncServiceBase {
     if (initialData.length === 0) {
       throw new Error(`No data exists for collection [${flow_name}], cannot initialise`);
     }
-    const schema = this.inferSchema(initialData[0]);
+    // add index property to each element before insert, for sorting queried data by original order
+    const initialDataWithMeta = initialData.map((el) => {
+      return {
+        ...el,
+        index_original: initialData.indexOf(el),
```

Comment:
Originally I was going to name this field "_index", however RxJs throws an error when trying to use a field name that starts with an underscore: `RxError: fieldnames do not match the regex:"^[a-zA-Z](?:[[a-zA-Z0-9_]*]?[a-zA-Z0-9])?$"`. Using "_index" might also get it confused with the `_index` of an item in `itemContext`, which is the index within the array of items to render, after any operations have been performed on the original data.

So it's an open question as to what this field could be named. It should be something that we're not going to overwrite as a property of the item. With #2310 in mind, we could prefix protected properties with a custom prefix: e.g., in this case we call the field something like `PROTECTED_index`, and as a solution to #2310, we append any field name that starts with `_` with the prefix "protected", e.g. `PROTECTED_translations`. Thoughts?